### DISCUSSION TITLE
Qtouch with megaTinyCore?

### DISCUSSION THREAD

**Discussion by grandaspanna** [SpenceKonde/megaTinyCore#368] [Q&A]
Does anybody have any experience using the QTouch library with the core? I have a project in mind for a 1614 and have tested the QTouch library on its own in Studio, but now want to expand to use some other libraries in Arduino IDE. Is there a "best practice" ?

FYI, for those not familiar, the PTC (peripheral touch controller) is pretty much a black box from a Microchip perspective and only binary versions of QTouch are readily available.

**Comment by SpenceKonde** [comment]
There is another issue (I may have3 closed it after it went over a year with no action) taking about this. That other core, while missing many features and only supported the 3216 and 3217... They had a way to package the qtouch libraries - I think they used the Atmel or Microchip tooling to do it, but they had a few libraries and some header files defining a whole boatload of functions, that were in that precompiled binary... 
You'd be a hero if you could figure out how they did it and bring it to megaTinyCore (generating the files like they have, only for as many parts as possible should be sufficient - I can figure out how they made the compiler make use of it)

It strikes me that, starting from files like that, an individual with no respect for international copyright law, located somewhere beyond the reach of western legal systems, and who had nothing to lose from infuriating microchip (in other words someone who is not me - I may have no respect for copyrights, but I do fear their enforcement, and in any event, I prefer to stay on Microchip's good side and doing this is not the way to stay there)..... one could extract the .o files from the .a's with avr-ar -x, then dump them to assembly with avr-objdump -d -r - that would get you dissembled object files....  with the destination of the jumps and calls included....

A trivial python script to turn things like 
  98:   00 f4           brcc    .+0             ; 0x9a <check_for_aks_block+0x9a>
                        98: R_AVR_7_PCREL       .text.check_for_aks_block+0xb2

into 
  98:   00 f4           brcc    .+18
  

You'd need to figure out how to deal with calls (I don't know offhand) too, but you'd then end up with a legal (syntactically - as I read the license, this is quite certainly not legal to do - or at least to distribute if you did) assembly listing. which could then be turned into (again, probsbly easiest to do wiyh python script) into a bunch of file declarations full of assembly. They could be declared ((naked)) since these definitions alreadty have the prologue and epilogue.... and then pasted into your new arduino library, a working library could be made that way...

But yeah, that's pretty much a non-starter thanks to Atmel -> microchip's mystefying decision to keep QTouch on lockdown, even if it costs them sales, which I am sure it has. You *really* wonder what they are hiding.... Would it have revealed too much about the internals of the design? Did they keep it secret just because they didn't trust the users to write code that worked, and the license was just incidental/reflexive secrecy? Was there something about the code that reflected poorly on the author? Maybe they used an algorithm that they' would be required to pay license fees for and they're not?  Could be anything! AAaanyway..... 

What that other core did doesn't seem to have gotten them in trouble, so I think we could do that. I just don't care about qtouch, and have too much else to do.

**Reply by grandaspanna** [reply]
I've made some promising progress, but hit a small issue with what looks like Arduino configuration. The steps I've taken:
1) Extract each of the .o files and combined them into a single .a
2) Created a new library folder, popped the header files into the src subfolder and created a library.properties file that includes the directive "precompiled=true". The .a file goes into a subfolder which I think should be "attinty1614"
3) Made two small changes to platform.txt. A) include {compiler.libraries.ldflags} into the "recipe.c.combine.pattern=" line, just after {objects} B) Added a new line "compiler.libraries.ldflags="
4) All goes great until the linking and something is not quite right, so I get this message:
Compiling libraries...
Compiling library "QTouch"
Library QTouch has been declared precompiled:
Precompiled library in "C:\Users\heyyou\Documents\Arduino\libraries\QTouch\src\attiny{build.attiny}" not found

How do I get Arduino to turn {build.attiny} into "1614"?

**Reply by grandaspanna** [reply]
FWIW, this is the best summary of how to set up precompiled libraries:
[https://forum.pjrc.com/threads/63256-Solution-for-adding-support-for-pre-compiled-Libraries](https://forum.pjrc.com/threads/63256-Solution-for-adding-support-for-pre-compiled-Libraries)

**Reply by grandaspanna** [reply]
Ok, I hacked boards.txt and I can get it to compile and link. This is clearly a hack:
atxy4.build.mcu=attiny1614
#atxy4.build.mcu=attiny{build.attiny}

Is this supposed to resolve automatically?

Now to spend some time seeing if I can actually get it to work!

**Reply by SpenceKonde** [reply]
interesting - so it wants to find it in /src/{build.mcu} - and substitutes that in, but either doesn[t pull in all the other variables it normally has, or just doesn't bother doing a check for things to substitute into build.mcu thinking nobody would ever do what I did there,,,,

Sounds like you are making excellent progress! it would be awesome if we could add touch support here :-)

I will fix boards.txt to give it what it needs. a while back I needed a  build variable with just the number of the part to make better names for exported assets (hex files and from next version onward, memory maps), and then I was like,,, hey instead of havign build.mcu defined for every chip, i could just use the build.attiny define I just added and append it to attiny!

**Reply by SpenceKonde** [reply]
This is sorted in master BUT you need to use a type B timer for millis until I can fix this new bug. Micros has backwards timetravel and has for a while, b ut the new delay() implementation cant' deal with that.

**Reply by SpenceKonde** [reply]
I spent a MONTH workibg on this last year, there was NO  backwards time travel in tghe whole core. When did this garbage get back in???

**Reply by SpenceKonde** [reply]
apparently, within 2 versions of my fixing the bugs. They were fixed in 1.1.9, 1.1.10 came next, then 2.0.0 which had backwards time travel...
